include/npcomp/Dialect/RD/IR/RDOps.td

```diff
@@ -0,0 +1,109 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef RD_OPS
+#define RD_OPS
+
+include "npcomp/Dialect/RD/IR/RDBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class RD_Op<string mnemonic, list<OpTrait> traits = []>
+    : Op<RD_Dialect, mnemonic, traits> {
+}
+
+// TODO: Consider defining a region for representing Dataset construction
+// instead of relying on FuncOp's to enforce additional constraints.
+
+def RD_MakeIteratorOp : RD_Op<"make_iterator"> {
+    let summary = "Builds an iterator for a given dataset.";
+    let description = [{
+        Builds an iterator for a given dataset.
+    }];
+
+    let arguments = (ins FlatSymbolRefAttr:$ds, Variadic<AnyType>:$operands);
+    let results = (outs RD_Iterator:$itr);
+
+    // TODO: add verification to ensure that types match up!!
+
+    let assemblyFormat = [{
+        $ds `(` $operands `)` attr-dict `:` functional-type(operands, results)
+    }];
+}
+
+def RD_IteratorNextOp : RD_Op<"iterator_next"> {
+    let summary = "Advances the iterator by one and returns the current values.";
+    let description = [{
+        Returns a tuple of values:
+         1. A Boolean indicating if there was a valid value.
+         2. The value (only valid if the boolean is true).
+    }];
+
+    let arguments = (ins RD_Iterator:$iterator);
+    let results = (outs I1:$valid, I64:$value);  // TODO: make generic over values.
+
+    let assemblyFormat = "$iterator attr-dict `:` functional-type(operands, results)";  // TODO: just write the element type in future?
+}
+
+def RD_Print : RD_Op<"print"> {
+    let summary = "Prints to the console.";
+    let description = [{
+        Example:
+
+        rd.print %value : i64
+    }];
+
+    let arguments = (ins I64:$value);
+
+    let assemblyFormat = "$value attr-dict `:` type($value)";
+}
+
+// Dataset ops allow users to build up compositions of computations.
+
+class RD_DatasetOp<string mnemonic>
+    : RD_Op<mnemonic, [NoSideEffect]> {
+}
+
+def RD_RangeOp : RD_DatasetOp<"range"> {
+    let summary = "Range of values.";
+    let description = [{
+        A sequence of consecutive integers.
+    }];
+
+    let arguments = (ins I64:$start, I64:$end, I64:$step);
+    let results = (outs RD_Dataset:$result);
+
+    let assemblyFormat = "$start `to` $end `by` $step attr-dict `:` functional-type(operands, results)";
+}
+
+def RD_InlineMapOp : RD_DatasetOp<"inline_map"> {
+    let summary = "Transform element values by applying function `f` to each of them.";
+    let description = [{
+        A serial, sequential execution implementation of the elementwise application of
+        f to each of the elements in this dataset.
+    }];
+
+    let arguments = (ins RD_Dataset:$src, FlatSymbolRefAttr:$f);
```

Comment:
I assume this is a symbol ref to a `func`, right? I suspect you will ulitmately want something more general than that (some kind of bound function/closure).

Reply:
Yes, I agree. So you know of a good closure MLIR implementation?

Reply:
I had one sketched for npcomp but not implemented. I suspect that most of the things you going to find are far too low level for what you want to be specifying here.

I had intended to extend the basicpy.slot_object in this direction and then constrain that a closure func was a regular func that took a slot_object as first argument. If you're going for representing python semantics, it will be quite involved for fill generality. I would probably start with a type that you could extend later and just reach out for now how to reference a non-closure/direct-call function symbolically.

Reply:
You could just have a variant of this op that takes a region.